File: src/it/java/quebec/virtualite/backend/utils/RestClient.java
package quebec.virtualite.backend.utils;

import io.restassured.RestAssured;
import io.restassured.response.Response;
import io.restassured.specification.RequestSpecification;
import org.springframework.stereotype.Component;

import static io.restassured.RestAssured.given;
import static io.restassured.http.ContentType.JSON;
import static org.apache.commons.lang3.StringUtils.isEmpty;

@Component
public class RestClient
{
    private static final char NON_BREAKING_SPACE = (char) 0x00A0;

    private Response response;

    private String username;
    private String password;

    public void connect(int serverPort)
    {
        RestAssured.port = serverPort;
        clearUser();
    }

    public void delete(String url, RestParam... params)
    {
        response = securedRequest()
            .delete(urlWithParams(url, params));
    }

    public void get(String url, RestParam... params)
    {
        response = securedRequest()
            .get(urlWithParams(url, params));
    }

    public void login(String username, String password)
    {
        this.username = username;
        this.password = password;
    }

    public void logout()
    {
        clearUser();
    }

    public void post(String url, Object dto, RestParam... params)
    {
        response = securedRequest()
            .contentType(JSON)
            .body(dto)
            .post(urlWithParams(url, params));
    }

    public void put(String url, Object dto)
    {
        response = securedRequest()
            .contentType(JSON)
            .body(dto)
            .put(url);
    }

    public Response response()
    {
        return response;
    }

    public String trim(String json)
    {
        return json
            .replace(" ", "")
            .replace("\r", "")
            .replace("\n", "")
            .replace(NON_BREAKING_SPACE, ' ');
    }

    protected String urlWithParams(String url, RestParam[] params)
    {
        for (RestParam param : params)
        {
            url = setParam(url, param);
        }

        return url;
    }

    private void clearUser()
    {
        username = "";
        password = "";
    }

    private boolean notIsLoggedIn()
    {
        return isEmpty(username) || isEmpty(password);
    }

    private RequestSpecification securedRequest()
    {
        if (notIsLoggedIn())
            return given();

        return given()
            .auth()
            .basic(username, password);
    }

    private String setParam(String url, RestParam param)
    {
        String paramName = "{" + param.key + "}";
        String paramValue = String.valueOf(param.value);
        char separator = url.contains("?")
                         ? '&'
                         : '?';

        return url.contains(paramName)
               ? url.replace(paramName, paramValue)
               : String.format("%s%c%s=%s", url, separator, param.key, paramValue);
    }
}

File: src/it/java/quebec/virtualite/backend/utils/RestClient.kt
package quebec.virtualite.backend.utils

import io.restassured.RestAssured
import io.restassured.http.ContentType
import io.restassured.response.Response
import io.restassured.specification.RequestSpecification
import org.apache.commons.lang3.StringUtils.isEmpty
import org.springframework.stereotype.Component

@Component
class RestClient
{
    private val NON_BREAKING_SPACE = 0x00A0.toChar()
    private val XSRF_TOKEN = "XSRF-TOKEN"
    private val X_XSRF_TOKEN = "X-XSRF-TOKEN"

    private lateinit var response: Response
    private lateinit var username: String
    private lateinit var password: String

    fun connect(serverPort: Int)
    {
        RestAssured.port = serverPort
        clearUser()
    }

    fun delete(url: String, vararg params: RestParam)
    {
        response = requestForWrites()
            .delete(urlWithParams(url, params))
    }

    fun get(url: String, vararg params: RestParam)
    {
        response = requestForReads()[urlWithParams(url, params)]
    }

    fun login(username: String, password: String)
    {
        this.username = username
        this.password = password
    }

    fun logout()
    {
        clearUser()
    }

    fun <T> post(url: String, payload: T, vararg params: RestParam)
    {
        response = requestForWrites()
            .contentType(ContentType.JSON)
            .body(payload)
            .post(urlWithParams(url, params))
    }

    fun <T> put(url: String, payload: T)
    {
        response = requestForWrites()
            .contentType(ContentType.JSON)
            .body(payload)
            .put(url)
    }

    fun response(): Response
    {
        return response
    }

    fun trim(json: String): String
    {
        return json
            .replace(" ", "")
            .replace("\r", "")
            .replace("\n", "")
            .replace(NON_BREAKING_SPACE, ' ')
    }

    fun urlWithParams(url: String, params: Array<out RestParam>): String
    {
        var urlWithParam = url
        for (param in params)
        {
            urlWithParam = setParam(urlWithParam, param)
        }
        return urlWithParam
    }

    private fun clearUser()
    {
        username = ""
        password = ""
    }

    private val jSessionID: String
        get() = RestAssured.given()
            .auth().basic(username, password)["/user"]
            .sessionId

    private fun getToken(jSessionID: String): String
    {
        return RestAssured.given()
            .sessionId(jSessionID)
            .contentType(ContentType.JSON)["/user"]
            .cookie(XSRF_TOKEN)
    }

    private fun notIsLoggedIn(): Boolean
    {
        return isEmpty(username) || isEmpty(password)
    }

    private fun requestForReads(): RequestSpecification
    {
        return if (notIsLoggedIn()) RestAssured.given() else RestAssured.given()
            .auth()
            .basic(username, password)
    }

    private fun requestForWrites(): RequestSpecification
    {
        if (notIsLoggedIn())
            return RestAssured.given()

        val jSessionID = jSessionID
        val token = getToken(jSessionID)
        return RestAssured.given()
            .sessionId(jSessionID)
            .header(X_XSRF_TOKEN, token)
    }

    private fun setParam(url: String, param: RestParam): String
    {
        val paramName = "{" + param.key + "}"
        return when
        {
            url.contains(paramName) -> url.replace(paramName, "${param.value}")

            else -> url +
                "${if (url.contains('?')) "&" else "?"}${param.key}=${param.value}"
        }
    }
}
